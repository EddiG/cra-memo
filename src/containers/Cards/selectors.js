import get from 'lodash/get';

export const selectCards = state => {
  const allIds = get(state, 'cards.allIds', []);
  return allIds.reduce((acc, id) => {
    const card = get(state, `cards.byId.${id}`);
    if (card) {
      const frontImage = get(
        state,
        `frontImages.byId.${card.frontImageId}.src`,
        '',
      );
      const backImage = get(
        state,
        `backImages.byId.${card.backImageId}.src`,
        '',
      );
      return [
        ...acc,
        {
          id: card.id,
          isOpened: card.isOpened,
          isGuessed: card.isGuessed,
          frontImage,
          backImage,
        },
      ];
    }
    return acc;
  }, []);
};

export const selectOpenedNotGuessedCards = state =>
  selectCards(state).filter(card => card.isOpened && !card.isGuessed);
