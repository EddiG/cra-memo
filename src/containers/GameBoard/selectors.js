export const selectCardsQuantity = state => state.settings.cardsQuantity;
export const selectFrontImages = state => state.frontImages.allIds;
export const selectBackImages = state => state.backImages.allIds;
