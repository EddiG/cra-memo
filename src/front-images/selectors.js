import toArray from 'lodash/toArray';

export const selectImagesSources = state =>
  toArray(state.byId).map(image => image.src);

export default selectImagesSources;
