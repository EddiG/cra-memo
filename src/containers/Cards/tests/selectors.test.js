import { selectCards, selectOpenedNotGuessedCards } from '../selectors';

const state = {
  cards: {
    byId: {
      card1: {
        id: 'card1',
        isOpened: true,
        isGuessed: false,
        frontImageId: 'frontImage1',
        backImageId: 'backImage1',
      },
      card2: {
        id: 'card2',
        isOpened: true,
        isGuessed: true,
        frontImageId: 'frontImage2',
        backImageId: 'backImage2',
      },
    },
    allIds: ['card1', 'card2'],
  },
  frontImages: {
    byId: {
      frontImage1: {
        id: 'frontImage1',
        src: 'source1',
      },
      frontImage2: {
        id: 'frontImage2',
        src: 'source2',
      },
    },
    allIds: ['frontImage1', 'frontImage2'],
  },
  backImages: {
    byId: {
      backImage1: {
        id: 'backImage1',
        src: 'source1',
      },
      backImage2: {
        id: 'backImage2',
        src: 'source2',
      },
    },
    allIds: ['backImage1', 'backImage2'],
  },
  settings: {
    cardsQuantity: 4,
  },
};
describe('Cards selector', () => {
  describe('selectCards', () => {
    it('should return the array of cards from the state', () => {
      expect(selectCards(state)).toMatchSnapshot();
    });
  });
  describe('selectOpenedNotGuessedCards', () => {
    it('should return the array of opened cards, except guessed', () => {
      expect(selectOpenedNotGuessedCards(state)).toMatchSnapshot();
    });
  });
});
