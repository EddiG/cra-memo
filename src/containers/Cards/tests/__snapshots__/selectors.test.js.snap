// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cards selector selectCards should return the array of cards from the state 1`] = `
Array [
  Object {
    "backImage": "source1",
    "frontImage": "source1",
    "id": "card1",
    "isGuessed": false,
    "isOpened": true,
  },
  Object {
    "backImage": "source2",
    "frontImage": "source2",
    "id": "card2",
    "isGuessed": true,
    "isOpened": true,
  },
]
`;

exports[`Cards selector selectOpenedNotGuessedCards should return the array of opened cards, except guessed 1`] = `
Array [
  Object {
    "backImage": "source1",
    "frontImage": "source1",
    "id": "card1",
    "isGuessed": false,
    "isOpened": true,
  },
]
`;
